Blog Post: double-checked locking (#78)
sampsyo/cs6120: content/blog/2019-10-18-double-checked-locking/index.md

    public Helper getHelper() {
        if (helper == null) {
            Helper h;
            synchronized(this) {
                h = helper;
                if (h == null)
                    synchronized(this) {
                        h = new Helper();
                    }                       
                helper = h;
            }
        }
        return helper;
    }
}

```

The intuitive idea is that the lock releasing would act as a memory barrier,
so that `helper=h` will not be executed until the initialization 
in the synchronized section is done.

Unfortunately, the lock releasing is a one-way memory barrier on many processors.
It only enforces that the instructions in the synchronized section 
must be executed before lock releasing.
The instruction `helper=h` behind the memory barrier could still be moved into 
synchronized section and executed before the object initialization is done.

## Working Solutions

### Volatile in JDK5

Java introduced "volatile" since JDK5. 
A "volatile" makes the variable be stored and accessed in main memory.
Every read of a volatile is from main memory.
Every write of a volatile is to main memory.

> The "volatile" also provides a "happen-before" guarantee:

Reviewer: Hmm, I don't think I'd use the term "happens-before"… it's too intimately tied to the Lamport ordering notion. https://lamport.azurewebsites.net/pubs/time-clocks.pdf

* The reads/writes of other variables before a write to a volatile variable 
cannot be reordered after the write to the volatile variable.
* The reads/writes of other variables after a read from a volatile variable
cannot be reordered before the read from the volatile variable.

With this new feature, the double-checked locking issue is resolved by simply changing
the `helper` to volatile, since it guarantees that the initialization of the
object `new Helper()` happens before the assignment to the `helper`.

```Java
class Foo {
    private volatile Helper helper = null;
    public Helper getHelper() {
        if (helper == null) {
            synchronized(this) {    
                if (helper == null)
                    helper = new Helper();
            }
        }
        return helper;
    }
}
```

However, since all read and write operations of a volatile variable need to

> access main memory, which can be very slow. 

Reviewer: Again, it's not just that they "access main memory"—they also use special hardware support for atomic/synchronized accesses (i.e., the compiler inserts barriers anyway), so nothing is really gained.

An improvement can be done with a local variable, to reduce number of times 
accessing main memory.

```Java
class Foo {
    private volatile Helper helper = null;
    public Helper getHelper() {
        Helper h = helper;
        if (h == null) {
            synchronized(this) {    
                h = helper;
                if (h == null) {
                    h = new Helper();
                    helper = h;
                }
            }
        }
        return h;
    }
}

> 

Reviewer: Extra blank line.

```

In cases that the `helper` in already initialized, 
   this optimization can reduce one volatile read by returning the local variable.

### Explicit Memory Barrier

The last fix with two synchronized sections does not work because releasing lock

> is an implicit "one-way" memory barrier.

Reviewer: Do you mean the _first_ fix? I think the _last_ fix, just above, actually does work…

```C++
class Foo {
    private:
        std::atomic <Foo*> helper;
    public:
        Foo* get_helper() {
            Foo* h = helper.load(std::memory_order_relaxed);
            std::atomic_thread_fence(std::memory_order_acquire);        //memory barrier
            if (h == nullptr) {
                std::lock_guard<std::mutex> lock(m_init);
                h = helper.load(std::memory_order_relaxed);
                if (h == nullptr) {
                    h = new Helper;
                    std::atomic_thread_fence(std::memory_order_release);//memory barrier
                    helper.store(h, std::memory_order_relaxed);
                }
            }
            return tmp;
        }
};
```

* `std::atomic_thread_fence(std::memory_order_acquire)` guarantees that 
read/write after a memory barrier cannot be reordered before the memory barrier.
* `std::atomic_thread_fence(std::memory_order_release)` guarantees that 
read/write before a memory barrier cannot be reordered after the memory barrier.

Similar to the Java volatile case, 
the initialization must be done before it stores `h` to `helper` 
with the memory barrier.

### Static Singleton

> If the `helper` is static, i.e. all the instances of class `Foo` shares the 

Reviewer: ```suggestion
If the `helper` is static, i.e., all the instances of class `Foo` share the 
```
https://capra.cs.cornell.edu/styleguide/#egie

same instance of `helper`, defining the `helper` in a static field of a separate
class will solve the problem.

```Java
class Foo {
    private static class HelperSingleton {
        public static final Helper helper = new Helper();
    }

    public Helper getHelper() {
        return HelperSingleton.helper;
    }
}
```

This is known as **initialization-on-demand holder idiom**, 
which is a safe and efficient concurrent lazy initialization 
for all Java versions.

* **Lazy Initialization**: as specified by the Java language,
    the static class `HelperSingleton` is not initialized until the first it is
    referenced.
* **Safe**: the class initialization is guaranteed to be synchronized by 
    Java language.
* **Efficient**: all subsequent calls of `getHelper` will return the initialized

>     `helper` without any synchronization overhead.

Reviewer: Wow—I'm not sure I understand this deeply Java-specific trick. You say that class initialization is guaranteed to be "synchronized," but synchronized _with what_?

That is, regardless of how it goes about it, a safe version of this will need to somehow use hardware synchronization to order the initialization with the use. So it does not suffice just to have all initializations be synchronized _with each other_. How does the JVM avoid acquiring a lock on subsequent accesses to `HelperSingleton.helper`?

### Thread Local

`ThreadLocal` is a variable that each thread will have its own copy of its value.
Each thread can only access and modify its own copy of a thread local variable
independently of other threads.

A thread local can be used to maintain the state of "whether the state has gone
through the synchronized initialization". 

> Since if a thread has gone through the synchronized initialized once,

Reviewer: ```suggestion
If a thread has gone through the synchronized initialization once,
```

> it can be confident about that object is already initialized.

Reviewer: ```suggestion
it can be confident that that object is already initialized.
```

Inside the synchronized initialization section, 
only the first thread will find the object is `null` and initialized object.
All threads will change their per-thread state, 

>     so that they will not enter the synchronized section again.

Reviewer: Maybe it's worth expanding on the fact that _every_ thread will then execute the critical section _exactly once_? That's clearly more costly than the "ideal" double-checked locking pattern, if it were to work. But it's not much worse, so that's probably fine.

```Java
class Foo {
    private static ThreadLocal perThreadState = new ThreadLocal();
    private Helper helper = null;
    public Helper getHelper() {
        if (perThreadState.get() == null) {
            synchronized {
                if (helper == null)
                    helper = new Helper();
                perThreadState.set(perThreadState);
            }
        }
        return helper;
    }
}
```

### 32-bit Primitives

If the initialized value is a 32-bit primitive value, such as `int` or `float`,

>    the double-checked locking will work.

Author: Maybe I just don't include this in my post, I don't quite like/understand it either...

            }
        }
        return helper;
    }
}
```

The assignment to 32-bit primitives can only happen once the data is available.
And the assignment is atomic. 
Other threads will only see a ready-to-use value or 0,
      there is no intermediate state of "initializing".

In fact, if the function `TheMagicNumber` always returns the same value without
side effects, the locking is not required.

```Java
class Foo {
    private int magicNumber = 0;
    public int getMagicNumber() {
        if (magicNumber == 0) 
            magicNumber = TheMagicNumber();
        return helper;
    }
}
```

## Conclusion

The article discussed the problem of double-checked locking for lazy 
initialization in multi-threaded environments.
It analyzed why some intuitive solutions do not work,
   and also analyzed some working solutions.

Writing multi-threaded program is hard. 
Writing correct and safe multi-threaded program is even harder.
When analyzing the correctness of multi-threaded programs,
it requires the considerations of multiple components, 
including compilers, systems, and processors.
On the other hand, when designing compilers, systems, or processors,

> it also needs to take consideration of common used design pattern.

Reviewer: ```suggestion
one also needs to take into consideration commonly used design patterns.
```

class Foo {
    private Helper helper = null;
    public Helper getHelper() {
        synchronized(this) {
            if (helper == null)
                helper = new Helper();
            return helper;
        }
    }
}
```

<img src="lazy-init-with-lock.png" style="width: 100%;">

However, we only need this section of code to be synchronized for the first time.
After the object is created, acquiring and releasing lock is unnecessary,
and they can have a huge performance impact.

What we want is like the following:

<img src="lazy-init-ideal.png" style="width: 100%;">

Only the first thread will enter the synchronized section and create the object. 
Once the `helper` is initialized, all subsequent accesses can run in parallel
without synchronization.

Intuitively, we may come up with the following sequence to do this job:

1. Check if the object is initialized without locking.
If it is, then return the object immediately.
2. Acquire the lock and check again if the object is initialized.
If another thread has grabbed the lock before, 
   the current thread can see the object is created, and return the object.
3. Otherwise, the current thread will create the object and return.

> With the guideline above, we will get a code like following:

Reviewer: ```suggestion
With the guidelines above, we will get code like the following:
```

    public Helper getHelper() {
        if (helper == null) {              // first check
            synchronized(this) {
                if (helper == null)         // second check
                    helper = new Helper();
            }
        }
        return helper;
    }
}
```

This tactic is call double-checked locking.

### Double-checked locking is broken 

However, this code is not guaranteed to work.

`helper = new Helper()` is not an atomic operation,
it consists multiple instructions allocating space, initializing fields of the object,
and assigning address to `helper`. 

In order to show what is really happening there,
we expand `helper = new Helper()` with some pseudocode 
and inline the object initialization code.

```Java
class Foo {
    private Helper helper = null;
    public Helper getHelper() {
        if (helper == null) {
            synchronized(this) {
                if (helper == null) {
                    ptr = allocate();
                    ptr.field1 = initField1();
                    ptr.field2 = initField2();
                    helper = ptr;
                }
            }
        }
        return helper;
    }
}

> ```

Reviewer: Inlining the initialization method is a really good idea! This makes it clearer.

In order to improve overall performance,
some compilers, memory systems, or processors may reorder the instructions,
like moving `helper = ptr` before initializing fields of the object.

```Java
if (helper == null) {

>     synchronized(this) {

Reviewer: Maybe include the "outer" `if` check here for consistency?

        if (helper == null) {
            ptr = allocate();
            helper = ptr;
            ptr.field1 = initField1();
            ptr.field2 = initField2();
        }
    }
}
return helper;
```

This reordering is legal, since there is no data dependency between `helper = ptr`
and the instructions for initializing fields. 

> However, this reordering could let other thread access uninitialized fields of the object.

Reviewer: ```suggestion
However, this reordering could let other threads access uninitialized fields of the object.
```

* `std::atomic_thread_fence(std::memory_order_release)` guarantees that 
read/write operations before a memory barrier cannot be reordered with
write operations after the memory barrier.

The memory barriers guarantee that 
1. `h` loads the value from `helper` before starting object initialization.
2. object initialization finishes before storing the value to `helper`. 

### Atomic Operation

An atomic operation will either happen completely, or it does not happen at all.
This is no intermediate state, so that the side effects of an atomic operation
will not be visible until the operation is complete.

In previous analysis, we have seen that `h = new Helper()` can be interleaved
because it is not an atomic operations.
If this operation is atomic, the double-checked locking will work.

#### 32-bit Primitive Variables

Read and write operations of most primitive variables 
(except `long` and `double` since they are 64-bit)are atomic.
If the initialized value is a 32-bit primitive variable,
   assignment to the variable will only happen once the data is available.
Since the write operation is atomic, 
other threads will either see a ready-to-use value or 0.

> There is no middle state of "initializing".

Author: Ah I see the point. Since it is not `private volatile int magicNumber`, but it somehow behaves like a volatile variable, so it needs some specific memory model support from Java.
